ICE Bad type error in iterator code

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.

http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/

-->

### Code

```rust
pub enum Option<T> {
    None,
    Some(T),
}

pub enum Result<T, E> {
    Ok(T),
    Err(E),
}

pub trait TryFrom<T> {
    /// The type returned in the event of a conversion error.
    type Error;

    /// Performs the conversion.
    fn try_from(value: T) -> Result<Self, Self::Error>;
}

pub trait From<T> {
    fn from(_: T) -> Self;
}

impl<T> From<T> for T {
    fn from(t: T) -> T {
        t
    }
}

impl<T, U> TryFrom<U> for T
where
    T: From<U>,
{
    type Error = !;

    fn try_from(value: U) -> Result<Self, Self::Error> {
        Result::Ok(T::from(value))
    }
}

pub fn test(n: usize) {
    let a = <usize>::try_from(n);
}

```

### Meta

- What version of Rust GCC were you using, git sha if possible. 87e7c9dedacd4d12636e6c626aab14bcecf686f2

### Error output

```
test.rs:44:45: note: fill_param_ty bounds_compatible: param T type usize
   44 |     fn add_usize(&self, n: usize) -> Option<Self> {
      |                                             ^~~~
rust1: note: unify_site id={80} expected={usize} expr={T}
rust1: note: unify::go ltype={usize} rtype={T}
test.rs:25:8: error: expected ‘usize’ got ‘T’
   23 | impl<T, U> TryFrom<U> for T
      |                           ~
   24 | where
   25 |     T: From<U>,
      |        ^~~~
......
   45 |         match <usize>::try_from(n) {
      |                ~~~~~
rust1: internal compiler error: in setup_associated_types, at rust/typecheck/rust-hir-trait-resolve.cc:506
```

<!--
Include a backtrace if available.
-->
<details><summary>Backtrace</summary>
  <p>
  
  ```
rust1: internal compiler error: in setup_associated_types, at rust/typecheck/rust-hir-trait-resolve.cc:506
0x14b41f6 Rust::Resolver::AssociatedImplTrait::setup_associated_types(Rust::TyTy::BaseType const*, Rust::TyTy::TypeBoundPredicate const&)
        ../../gccrs/gcc/rust/typecheck/rust-hir-trait-resolve.cc:506
0x15034de Rust::Resolver::TypeCheckExpr::resolve_segments(unsigned int, std::vector<Rust::HIR::PathExprSegment, std::allocator<Rust::HIR::PathExprSegment> >&, unsigned long, Rust::TyTy::BaseType*, Rust::Analysis::NodeMapping const&, Location)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-path.cc:396
0x1501c56 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::QualifiedPathInExpression&)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-path.cc:38
0x137bc95 Rust::HIR::QualifiedPathInExpression::accept_vis(Rust::HIR::HIRExpressionVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir.cc:3787
0x14d005b Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-expr.cc:42
0x14d0a61 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::CallExpr&)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-expr.cc:179
0x137e473 Rust::HIR::CallExpr::accept_vis(Rust::HIR::HIRExpressionVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir.cc:4891
0x14d005b Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-expr.cc:42
0x14d6471 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::MatchExpr&)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-expr.cc:1390
0x137e2eb Rust::HIR::MatchExpr::accept_vis(Rust::HIR::HIRExpressionVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir.cc:4849
0x14d005b Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-expr.cc:42
0x14d2762 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-expr.cc:561
0x137eb79 Rust::HIR::BlockExpr::accept_vis(Rust::HIR::HIRExpressionVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir.cc:5083
0x14d005b Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-expr.cc:42
0x14decf9 Rust::Resolver::TypeCheckImplItem::visit(Rust::HIR::Function&)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-implitem.cc:311
0x137ec21 Rust::HIR::Function::accept_vis(Rust::HIR::HIRImplVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir.cc:5101
0x14de12b Rust::Resolver::TypeCheckImplItem::Resolve(Rust::HIR::ImplBlock*, Rust::HIR::ImplItem*, Rust::TyTy::BaseType*, std::vector<Rust::TyTy::SubstitutionParamMapping, std::allocator<Rust::TyTy::SubstitutionParamMapping> >)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-implitem.cc:180
0x14bec02 Rust::Resolver::TypeCheckItem::visit(Rust::HIR::ImplBlock&)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-item.cc:341
0x137eeb7 Rust::HIR::ImplBlock::accept_vis(Rust::HIR::HIRVisItemVisitor&)
        ../../gccrs/gcc/rust/hir/tree/rust-hir.cc:5173
0x14bc906 Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
        ../../gccrs/gcc/rust/typecheck/rust-hir-type-check-item.cc:49
  ```
  
  </p>
</details>

--- a/gcc/rust/typecheck/rust-hir-trait-resolve.cc
+++ b/gcc/rust/typecheck/rust-hir-trait-resolve.cc
@@ -513,11 +513,9 @@ AssociatedImplTrait::setup_associated_types (
 	? SubstMapperInternal::Resolve (associated_self, infer_arguments)
 	: associated_self;
 
-  // FIXME this needs to do a lookup for the trait-reference DefId instead of
-  // assuming its the first one in the list
-  rust_assert (associated_self->num_specified_bounds () > 0);
-  TyTy::TypeBoundPredicate &impl_predicate
-    = associated_self->get_specified_bounds ().at (0);
+  const TyTy::TypeBoundPredicate &impl_predicate
+    = associated_self->lookup_predicate (bound.get_id ());
+  rust_assert (!impl_predicate.is_error ());
 
   // infer the arguments on the predicate
   std::vector<TyTy::BaseType *> impl_trait_predicate_args;

--- a/gcc/rust/typecheck/rust-hir-type-check-item.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-item.cc
@@ -73,6 +73,51 @@ TypeCheckItem::ResolveImplBlockSelf (HIR::ImplBlock &impl_block)
   return resolver.resolve_impl_block_self (impl_block);
 }
 
+TyTy::BaseType *
+TypeCheckItem::ResolveImplBlockSelfWithInference (HIR::ImplBlock &impl,
+						  Location locus)
+{
+  TypeCheckItem resolver;
+
+  bool failed_flag = false;
+  std::vector<TyTy::SubstitutionParamMapping> substitutions
+    = resolver.resolve_impl_block_substitutions (impl, failed_flag);
+  if (failed_flag)
+    {
+      return new TyTy::ErrorType (impl.get_mappings ().get_hirid ());
+    }
+
+  // now that we have the param mappings we need to query the self type
+  TyTy::BaseType *self = resolver.resolve_impl_block_self (impl);
+
+  // nothing to do
+  if (substitutions.empty () || self->is_concrete ())
+    return self;
+
+  // generate inference variables for the subst-param-mappings
+  std::vector<TyTy::SubstitutionArg> args;
+  for (auto &p : substitutions)
+    {
+      if (p.needs_substitution ())
+	{
+	  TyTy::TyVar infer_var = TyTy::TyVar::get_implicit_infer_var (locus);
+	  args.push_back (TyTy::SubstitutionArg (&p, infer_var.get_tyty ()));
+	}
+      else
+	{
+	  TyTy::ParamType *param = p.get_param_ty ();
+	  TyTy::BaseType *resolved = param->destructure ();
+	  args.push_back (TyTy::SubstitutionArg (&p, resolved));
+	}
+    }
+
+  // create argument mappings
+  TyTy::SubstitutionArgumentMappings infer_arguments (std::move (args), {},
+						      locus);
+
+  return SubstMapperInternal::Resolve (self, infer_arguments);
+}
+
 void
 TypeCheckItem::visit (HIR::TypeAlias &alias)
 {

--- a/gcc/rust/typecheck/rust-hir-type-check-item.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-item.cc
@@ -74,8 +74,9 @@ TypeCheckItem::ResolveImplBlockSelf (HIR::ImplBlock &impl_block)
 }
 
 TyTy::BaseType *
-TypeCheckItem::ResolveImplBlockSelfWithInference (HIR::ImplBlock &impl,
-						  Location locus)
+TypeCheckItem::ResolveImplBlockSelfWithInference (
+  HIR::ImplBlock &impl, Location locus,
+  TyTy::SubstitutionArgumentMappings *infer_arguments)
 {
   TypeCheckItem resolver;
 
@@ -112,10 +113,20 @@ TypeCheckItem::ResolveImplBlockSelfWithInference (HIR::ImplBlock &impl,
     }
 
   // create argument mappings
-  TyTy::SubstitutionArgumentMappings infer_arguments (std::move (args), {},
-						      locus);
+  *infer_arguments
+    = TyTy::SubstitutionArgumentMappings (std::move (args), {}, locus);
 
-  return SubstMapperInternal::Resolve (self, infer_arguments);
+  TyTy::BaseType *infer = SubstMapperInternal::Resolve (self, *infer_arguments);
+
+  // we only need to apply to the bounds manually on types which dont bind
+  // generics
+  if (!infer->has_subsititions_defined ())
+    {
+      for (auto &bound : infer->get_specified_bounds ())
+	bound.handle_substitions (*infer_arguments);
+    }
+
+  return infer;
 }
 
 void

--- a/gcc/rust/typecheck/rust-hir-type-check-item.h
+++ b/gcc/rust/typecheck/rust-hir-type-check-item.h
@@ -35,6 +35,9 @@ public:
 
   static TyTy::BaseType *ResolveImplBlockSelf (HIR::ImplBlock &impl_block);
 
+  static TyTy::BaseType *
+  ResolveImplBlockSelfWithInference (HIR::ImplBlock &impl, Location locus);
+
   void visit (HIR::Module &module) override;
   void visit (HIR::Function &function) override;
   void visit (HIR::TypeAlias &alias) override;

--- a/gcc/rust/typecheck/rust-hir-type-check-item.h
+++ b/gcc/rust/typecheck/rust-hir-type-check-item.h
@@ -35,8 +35,9 @@ public:
 
   static TyTy::BaseType *ResolveImplBlockSelf (HIR::ImplBlock &impl_block);
 
-  static TyTy::BaseType *
-  ResolveImplBlockSelfWithInference (HIR::ImplBlock &impl, Location locus);
+  static TyTy::BaseType *ResolveImplBlockSelfWithInference (
+    HIR::ImplBlock &impl, Location locus,
+    TyTy::SubstitutionArgumentMappings *infer_arguments);
 
   void visit (HIR::Module &module) override;
   void visit (HIR::Function &function) override;

--- a/gcc/rust/typecheck/rust-hir-type-check-path.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-path.cc
@@ -415,11 +415,8 @@ TypeCheckExpr::resolve_segments (NodeId root_resolved_node_id,
 	    = context->lookup_associated_trait_impl (impl_block_id,
 						     &associated);
 	  TyTy::BaseType *impl_block_ty
-	    = TypeCheckItem::ResolveImplBlockSelf (*associated_impl_block);
-
-	  if (impl_block_ty->needs_generic_substitutions ())
-	    impl_block_ty
-	      = SubstMapper::InferSubst (impl_block_ty, seg.get_locus ());
+	    = TypeCheckItem::ResolveImplBlockSelfWithInference (
+	      *associated_impl_block, seg.get_locus ());
 
 	  prev_segment = unify_site (seg.get_mappings ().get_hirid (),
 				     TyTy::TyWithLocation (prev_segment),

--- a/gcc/rust/typecheck/rust-hir-type-check-path.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-path.cc
@@ -431,23 +431,15 @@ TypeCheckExpr::resolve_segments (NodeId root_resolved_node_id,
 	      // we need to setup with apropriate bounds
 	      HIR::TypePath &bound_path
 		= *associated->get_impl_block ()->get_trait_ref ().get ();
+	      const auto &trait_ref = *TraitResolver::Resolve (bound_path);
+	      rust_assert (!trait_ref.is_error ());
 
-	      // generate an implicit HIR Type we can apply to the predicate
-	      HirId implicit_id = mappings->get_next_hir_id ();
-	      context->insert_implicit_type (implicit_id, impl_block_ty);
-
-	      Analysis::NodeMapping mappings (expr_mappings.get_crate_num (),
-					      expr_mappings.get_nodeid (),
-					      implicit_id,
-					      expr_mappings.get_local_defid ());
-	      HIR::TypePath *implicit_self_bound
-		= new HIR::TypePath (mappings, {},
-				     Linemap::predeclared_location (), false);
-
-	      TyTy::TypeBoundPredicate predicate
-		= get_predicate_from_bound (bound_path, implicit_self_bound);
-	      impl_block_ty
-		= associated->setup_associated_types (prev_segment, predicate);
+	      const auto &predicate
+		= impl_block_ty->lookup_predicate (trait_ref.get_defid ());
+	      if (!predicate.is_error ())
+		impl_block_ty
+		  = associated->setup_associated_types (prev_segment,
+							predicate);
 	    }
 	}
 

--- a/gcc/rust/typecheck/rust-hir-type-check-path.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-path.cc
@@ -414,9 +414,16 @@ TypeCheckExpr::resolve_segments (NodeId root_resolved_node_id,
 	  bool found_impl_trait
 	    = context->lookup_associated_trait_impl (impl_block_id,
 						     &associated);
+
+	  auto mappings = TyTy::SubstitutionArgumentMappings::error ();
 	  TyTy::BaseType *impl_block_ty
 	    = TypeCheckItem::ResolveImplBlockSelfWithInference (
-	      *associated_impl_block, seg.get_locus ());
+	      *associated_impl_block, seg.get_locus (), &mappings);
+
+	  // we need to apply the arguments to the segment type so they get
+	  // unified properly
+	  if (!mappings.is_error ())
+	    tyseg = SubstMapperInternal::Resolve (tyseg, mappings);
 
 	  prev_segment = unify_site (seg.get_mappings ().get_hirid (),
 				     TyTy::TyWithLocation (prev_segment),

--- a/gcc/rust/typecheck/rust-hir-type-check-path.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-path.cc
@@ -427,9 +427,7 @@ TypeCheckExpr::resolve_segments (NodeId root_resolved_node_id,
 				     seg.get_locus ());
 	  bool ok = prev_segment->get_kind () != TyTy::TypeKind::ERROR;
 	  if (!ok)
-	    {
-	      return;
-	    }
+	    return;
 
 	  if (found_impl_trait)
 	    {

--- a/gcc/rust/typecheck/rust-tyty-subst.cc
+++ b/gcc/rust/typecheck/rust-tyty-subst.cc
@@ -168,10 +168,14 @@ SubstitutionParamMapping::override_context ()
 SubstitutionArg::SubstitutionArg (const SubstitutionParamMapping *param,
 				  BaseType *argument)
   : param (param), argument (argument)
-{}
+{
+  if (param != nullptr)
+    original_param = param->get_param_ty ();
+}
 
 SubstitutionArg::SubstitutionArg (const SubstitutionArg &other)
-  : param (other.param), argument (other.argument)
+  : param (other.param), original_param (other.original_param),
+    argument (other.argument)
 {}
 
 SubstitutionArg &
@@ -179,6 +183,8 @@ SubstitutionArg::operator= (const SubstitutionArg &other)
 {
   param = other.param;
   argument = other.argument;
+  original_param = other.original_param;
+
   return *this;
 }
 
@@ -198,6 +204,12 @@ const SubstitutionParamMapping *
 SubstitutionArg::get_param_mapping () const
 {
   return param;
+}
+
+const ParamType *
+SubstitutionArg::get_param_ty () const
+{
+  return original_param;
 }
 
 SubstitutionArg
@@ -227,7 +239,7 @@ SubstitutionArg::is_conrete () const
 std::string
 SubstitutionArg::as_string () const
 {
-  return param->as_string ()
+  return original_param->as_string ()
 	 + (argument != nullptr ? ":" + argument->as_string () : "");
 }
 
@@ -289,9 +301,7 @@ SubstitutionArgumentMappings::get_argument_for_symbol (
 {
   for (auto &mapping : mappings)
     {
-      const SubstitutionParamMapping *param = mapping.get_param_mapping ();
-      const ParamType *p = param->get_param_ty ();
-
+      const ParamType *p = mapping.get_param_ty ();
       if (p->get_symbol ().compare (param_to_find->get_symbol ()) == 0)
 	{
 	  *argument = mapping;

--- a/gcc/rust/typecheck/rust-tyty-subst.h
+++ b/gcc/rust/typecheck/rust-tyty-subst.h
@@ -87,6 +87,8 @@ public:
 
   const SubstitutionParamMapping *get_param_mapping () const;
 
+  const ParamType *get_param_ty () const;
+
   static SubstitutionArg error ();
 
   bool is_error () const;
@@ -97,6 +99,7 @@ public:
 
 private:
   const SubstitutionParamMapping *param;
+  const ParamType *original_param;
   BaseType *argument;
 };
 

--- a/gcc/testsuite/rust/compile/issue-1893.rs
+++ b/gcc/testsuite/rust/compile/issue-1893.rs
@@ -1,4 +1,3 @@
-// { dg-additional-options "-frust-compile-until=nameresolution" }
 pub enum Option<T> {
     None,
     Some(T),
@@ -10,10 +9,8 @@ pub enum Result<T, E> {
 }
 
 pub trait TryFrom<T> {
-    /// The type returned in the event of a conversion error.
     type Error;
 
-    /// Performs the conversion.
     fn try_from(value: T) -> Result<Self, Self::Error>;
 }
 

